File: LargestRectangleHistogram.java
// logic -
// montonically increasingly stack used. 
// so when a lower height is encountered, the previous heights are removed 
//till there are other heights which are lower or equal to this height in the stack

// time complexity = O(n)
// space complexity = O(n)
class Solution {
    public int largestRectangleArea(int[] heights) {
        int max=0;
        int n = heights.length;
        Stack<Integer> stack = new Stack<>();
        stack.push(-1); // left most boundary
        for(int i=0;i< n ;i++) {
            while(stack.peek() != -1 && heights[i] < heights[stack.peek()]) {
                int popped = stack.pop();
                int currArea = heights[popped] * (i-stack.peek()-1); // -1 is added for the fact that the histogram width is 1.
                max = Math.max(max, currArea);
            }
            stack.push(i);
        }
        // end of the stack.
        while(stack.peek() !=-1) {
            int currArea = heights[stack.pop()] * (n-stack.peek()-1);
            max = Math.max(max, currArea);
        }
        
        return max;
    }
}

File: Largest Rectangle_Histogram.py
# Approach:
# - We use a **monotonic stack** to efficiently compute the largest rectangle area.
# - The key idea is to determine the **width** of the rectangle using the nearest smaller bar to the left and right.
# - We iterate through `heights` and maintain a stack storing indices of increasing height bars.
# - If we encounter a smaller bar, we **pop from the stack** and compute the area using the popped height as the smallest bar.
# - At the end, we process any remaining bars in the stack to compute their maximum possible rectangle area.

# Time Complexity: O(N) - Each bar is pushed and popped from the stack once.
# Space Complexity: O(N) - Stack stores at most N elements in the worst case.

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []  # Stack to store indices of increasing height bars
        max_area = 0  # Variable to track the maximum rectangle area
        n = len(heights)

        for i in range(n):
            # Maintain an increasing stack; process bars when a smaller bar is encountered
            while stack and heights[i] < heights[stack[-1]]:
                height = heights[stack.pop()]  # Pop the last bar (smallest height in current window)
                width = i if not stack else i - stack[-1] - 1  # Compute width using next smaller element in stack
                max_area = max(max_area, height * width)  # Update max area
            
            stack.append(i)  # Push the current index to stack

        # Process remaining elements in stack (bars extending to the end)
        while stack:
            height = heights[stack.pop()]
            width = n if not stack else n - stack[-1] - 1
            max_area = max(max_area, height * width)

        return max_area  # Return the largest rectangle area found
